### PR TITLE
Upgrade fields to store boundary conditions

### LINKED ISSUE #606: Storing boundary conditions inside fields
The issue of whether fields should store their boundary conditions was brought up in #601. We've discussed this idea before but I'm opening this issue to see if this is something we want to pursue.

Note: Moving boundary conditions to fields should happen alongside or after the introduce of grid topologies (#489) which would decouple the grid topology from the boundary conditions (although some boundary conditions will be determined by the grid topology).

I think it's a good idea that will pay off in code clarity and ease of use.
1. Functions such as [`fill_halo_regions!`](https://github.com/climate-machine/Oceananigans.jl/blob/8d1b164881f12607748bba044afb93c71c652fac/src/BoundaryConditions/fill_halo_regions.jl#L22) will simplify as we no longer need to pass in both a field and boundary conditions, just the field itself.
2. No need to prepare boundary condition tuples to pass to functions, for example, in [`complete_pressure_correction_step!`](https://github.com/climate-machine/Oceananigans.jl/blob/8d1b164881f12607748bba044afb93c71c652fac/src/TimeSteppers/TimeSteppers.jl#L177-L183).
3. Oceananigans has have a pretty deep hierarchy of boundary conditions: BC -> Coordinate BCs -> field BCs -> Solution BCs -> Model BCs. Moving BCs to fields will eliminate the need for solution BCs and model BCs, which will create a simpler hierarchy of just BC -> coordinate BCs -> field BCs. This will make it easier to developers and users to access and interact with boundary conditions, especially as [`ModelBoundaryConditions`](https://github.com/climate-machine/Oceananigans.jl/blob/master/src/BoundaryConditions/solution_and_model_boundary_conditions.jl#L155-L177) can only grow in complexity.
4. Storing fields in one struct and boundary conditions in another struct feels antithetical to orthogonal design.

Another reason to consider having fields carry their boundary conditions around is to avoid having to build more boundary condition machinery, i.e. more [solution_and_model_boundary_conditions.jl](https://github.com/climate-machine/Oceananigans.jl/blob/master/src/BoundaryConditions/solution_and_model_boundary_conditions.jl) when new fields are added or new models are added (e.g. #605).

User API may have to change as well, although I don't think we should get rid of the current `boundary_conditions` Model kwarg unless we have a better interface.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR (part 3/3) upgrades the field abstraction so fields store their own boundary conditions. This simplifies the model boundary condition hierarchy and generalizes the field and boundary conditions abstractions so they can be used for a compressible model (and any other model we come up with in the future). All future fields will have boundary conditions so PRs like #601 won't be necessary again.

The only change in user interface is that you pass a named tuple to the model constructor now instead of an instance of `SolutionBoundaryConditions`.

This also works for LES diffusivities so the amount of convoluted scripting gymnastics is much reduced (see test from #601). Setting a diffusivity BC is now almost as easy as a tracer BC.

```julia
grid = RegularCartesianGrid(FT, size=(16, 16, 16), length=(1, 1, 1))

buoyancy_bcs = TracerBoundaryConditions(grid, bottom=BoundaryCondition(Gradient, bz))
κₑ_bcs = DiffusivityBoundaryConditions(grid, bottom=BoundaryCondition(Value, κ₀))
model_bcs = (b=buoyancy_bcs, κₑ=(b=κₑ_bcs,))

model = IncompressibleModel(
    grid=grid, architecture=arch, float_type=FT, tracers=:b, buoyancy=BuoyancyTracer(),
    closure=AnisotropicMinimumDissipation(), boundary_conditions=model_bcs
)
```

Internally: No surprise, this change ended up being pretty invasive. But note that we now have a more flexible and easier to use package with fewer lines of code!

I'm happy to discuss and iterate over the choices that were made in this PR. But glad that I was able to make these changes. Development of the compressible model can continue based on this branch.

Changes:
1. Fields has a new property: `field.boundary_conditions`.

2. Better pretty printing for fields:
```
Field located at (Cell, Cell, Cell)
├── data: OffsetArrays.OffsetArray{Float64,3,Array{Float64,3}}, size: (18, 18, 18)
├── grid: RegularCartesianGrid{Float64, Periodic, Periodic, Bounded}(Nx=16, Ny=16, Nz=16)
└── boundary conditions: x=(west=Periodic, east=Periodic), y=(south=Periodic, north=Periodic), z=(bottom=NoFlux, top=NoFlux)
```

3. Field tuple constructors have two versions now. One that accepts boundary conditions (useful in model constructors) and another that accepts full fields (useful for checkpoint restoration and other custom functionality).

4. ~~You can pass anything into `VelocityFields` and `TracerFields`. Perhaps we should check whether everything passed into these functions is a `Field` or `AbstractField`. But for now that's extra code and it's not an exported function so I'm okay leaving it as is (we might want the flexibility in the future?).~~ A check was added in #627.

5. Model no longer has the `boundary_conditions` property.

6. There is no need for `SolutionBoundaryConditions` and `ModelBoundaryConditions` anymore: code is simpler :tada:

7. `TurbulentDiffusivities` has been renamed to `DiffusivityFields` for consistency with `VelocityFields` and `TracerFields`, etc.

8. For each turbulence closure, `DiffusivityFields` gets the same two versions (one if you want control over the full field, another if you just want to set boundary conditions).

9. Halo filling and parts of the time stepping have been simplified as there is no need to juggle around and bundle up boundary conditions.

Some comments:
1. It's not clear whether fields for abstract operations should have boundary conditions as they don't have halos and we don't impose boundary conditions on them. Do we want to generally set their boundary conditions to `nothing`? I did this for the abstract operations tests and they all passed.
2. We could further simplify the time stepping code if we adapt the `Field` abstraction to be GPU compatible (#298).
3. This is probably an edge case but I don't think setting diffusivity BCs when using a tuple of closures will actually work. Or rather, I'm not sure how to specify two different diffusivity BCs, one for each closure.
4. Before merging I still need to update the checkpointer and update the documentation.

Resolves #606

Note: This PR branches off from #628.

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
I started updating the checkpointer but didn't want to refactor too much as there's already an open PR that refactors it (#628). I'll wait until it's merged before I fully update the checkpointer here.

**Comment by ali-ramadhan** [comment]
With part 2 (#628) merged I merged master into this branch to resolve conflicts which also reduced the number of changes but it's still a pretty big and complicated PR.

@glwagner We should probably make sure that the changes and design choices in this PR make sense.

Once this PR is merged, I think we should release v0.23.0.

File: src/BoundaryConditions/field_boundary_conditions.jl [hunk 1]
```diff
@@ -83,5 +83,5 @@ end
   VVelocityBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:v; kwargs...)
   WVelocityBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:w; kwargs...)
      TracerBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:tracer; kwargs...)
-#   PressureBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:pressure, kwargs...)
-#DiffusivityBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:diffusivity, kwargs...)
+   PressureBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:pressure; kwargs...)
```
Comment (glwagner) [inline src/BoundaryConditions/field_boundary_conditions.jl]: what is the advantage of the "field_type" keyword? I think this is poor design because it will be harder to maintain and the code is not transparent. I would rather define these constructors here.

File: src/Fields/field.jl [hunk 2]
```diff
@@ -36,84 +36,90 @@ A type describing the location at the face of a grid cell.
...
+    Field(Cell, Cell, Face, arch, grid, bcs, data)
 
- CellField(T::DataType, arch, grid) = Field(Cell, Cell, Cell, arch, grid, zeros(T, arch, grid))
-XFaceField(T::DataType, arch, grid) = Field(Face, Cell, Cell, arch, grid, zeros(T, arch, grid))
-YFaceField(T::DataType, arch, grid) = Field(Cell, Face, Cell, arch, grid, zeros(T, arch, grid))
-ZFaceField(T::DataType, arch, grid) = Field(Cell, Cell, Face, arch, grid, zeros(T, arch, grid))
+ CellField(FT::DataType, arch, grid, bcs) =  CellField(arch, grid, bcs, zeros(FT, arch, grid))
+XFaceField(FT::DataType, arch, grid, bcs) = XFaceField(arch, grid, bcs, zeros(FT, arch, grid))
+YFaceField(FT::DataType, arch, grid, bcs) = YFaceField(arch, grid, bcs, zeros(FT, arch, grid))
+ZFaceField(FT::DataType, arch, grid, bcs) = ZFaceField(arch, grid, bcs, zeros(FT, arch, grid))
```
Comment (glwagner) [inline src/Fields/field.jl]: these constructors should be extended to default bcs derived from the topology of grid:

```julia
CellField(FT::DataType, arch, grid, bcs=TracerBoundaryConditions(grid)) = ...
```

et cetera.

File: src/Fields/field_tuples.jl [hunk 3]
```diff
@@ -1,30 +1,67 @@
...
 """
     TracerFields(arch, grid, tracer_names; kwargs...)
 
 Return a NamedTuple with tracer fields specified by `tracer_names` initialized as
-`CellField`s on the architecture `arch` and `grid`. Optional `kwargs` can be specified
-to assign data arrays to each tracer field.
+`CellField`s on the architecture `arch` and `grid`. Fields may be passed via optional
+keyword arguments `kwargs` for each field.
 """
 function TracerFields(arch, grid, tracer_names; kwargs...)
```
Comment (glwagner) [inline src/Fields/field_tuples.jl]: Even with the docstring I am a little confused how this function works. Maybe an example would help?

File: src/Fields/field_tuples.jl [hunk 4]
```diff
@@ -55,9 +108,9 @@ can be specified to assign data arrays to each tendency field.
 """
 function TendencyFields(arch, grid, tracer_names; kwargs...)
     velocities = (
-        u = :u ∈ keys(kwargs) ? XFaceField(arch, grid, kwargs[:u]) : XFaceField(arch, grid),
-        v = :v ∈ keys(kwargs) ? YFaceField(arch, grid, kwargs[:v]) : YFaceField(arch, grid),
-        w = :w ∈ keys(kwargs) ? ZFaceField(arch, grid, kwargs[:w]) : ZFaceField(arch, grid)
+        u = :u ∈ keys(kwargs) ? kwargs[:u] : XFaceField(arch, grid, UVelocityBoundaryConditions(grid), zeros(arch, grid)),
```
Comment (glwagner) [inline src/Fields/field_tuples.jl]: These can be defined in the constructor, avoiding the chain of if statements:

```julia
function TendencyFields(arch, grid, tracer_names; 
    u = XFaceField(arch, grid, UVelocityBoundaryConditions(grid), zeros(arch, grid),
    # similar for v and w
    tracer_kwargs...)
```

File: src/Models/incompressible_model.jl [hunk 5]
```diff
@@ -74,15 +73,15 @@ function IncompressibleModel(;
                coriolis = nothing,
           surface_waves = nothing,
                 forcing = ModelForcing(),
-    boundary_conditions = SolutionBoundaryConditions(grid),
+    boundary_conditions = NamedTuple(),
```
Comment (glwagner) [inline src/Models/incompressible_model.jl]: The grid is required `kwarg`, which means that we can at least infer default velocity boundary conditions here. Perhaps that would be better?

File: src/TimeSteppers/TimeSteppers.jl [hunk 6]
```diff
@@ -108,52 +123,32 @@ end
 Calculate the (nonhydrostatic) pressure correction associated `tendencies`, `velocities`, and step size `Δt`.
 """
 function calculate_pressure_correction!(nonhydrostatic_pressure, Δt, tendencies, velocities, model)
-    velocity_tendencies = (u=tendencies.u, v=tendencies.v, w=tendencies.w)
+    velocity_tendencies = (u=model.timestepper.Gⁿ.u, v=model.timestepper.Gⁿ.v, w=model.timestepper.Gⁿ.w)
```
Comment (glwagner) [inline src/TimeSteppers/TimeSteppers.jl]: This change means that `tendencies` is no longer needed in the function signature above?

File: src/TimeSteppers/TimeSteppers.jl [hunk 7]
```diff
@@ -92,12 +103,16 @@ contribution from non-hydrostatic pressure.
...
+    calculate_interior_source_terms!(
+        tendencies, model.architecture, model.grid, model.coriolis, model.buoyancy,
+        model.surface_waves, model.closure, velocities, tracers, pressures.pHY′,
+        diffusivities, model.forcing, model.parameters, model.clock.time
+    )
 
-    calculate_boundary_source_terms!(tendencies, model.boundary_conditions.solution, model.architecture,
-                                     model.grid, boundary_condition_function_arguments(model)...)
+    calculate_boundary_source_terms!(
+        model.timestepper.Gⁿ, model.architecture, model.velocities,
```
Comment (glwagner) [inline src/TimeSteppers/TimeSteppers.jl]: isnt `tendencies` the same as `model.timestepper.Gⁿ`? what is the intent of using `model.timestepper.Gⁿ` instead of tendencies? This function mixes them, which is unacceptable. We should either use `tendencies` or `model.timestepper.Gⁿ`, but not both.

File: src/TurbulenceClosures/diffusivity_fields.jl [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+function KappaFields(arch, grid, tracer_names; kwargs...)
+    κ_fields =
+        Tuple(c ∈ keys(kwargs) ?
+              kwargs[c] :
+              CellField(arch, grid, DiffusivityBoundaryConditions(grid), zeros(arch, grid))
+              for c in tracer_names)
+    return NamedTuple{tracer_names}(κ_fields)
+end
+
+function KappaFields(arch, grid, tracer_names, bcs::NamedTuple)
```
Comment (glwagner) [inline src/TurbulenceClosures/diffusivity_fields.jl]: could also use `TracerDiffusivityFields` and `ViscosityFields` --- it may be more explicit and clear.

File: src/Fields/field_tuples.jl (hunk 4 above)
Comment (glwagner) [reply src/Fields/field_tuples.jl]: Maybe I misinterpreted what is happening here. Either way the 3 if statements indicate that a better design is possible and should be pursued.

**Comment by glwagner** [comment]
> It's not clear whether fields for abstract operations should have boundary conditions as they don't have halos and we don't impose boundary conditions on them. Do we want to generally set their boundary conditions to nothing? I did this for the abstract operations tests and they all passed.

We should always use default boundary conditions for `grid` and the location, eg `Face` or `Cell` along each dimension.

**Comment by glwagner** [comment]
> We could further simplify the time stepping code if we adapt the Field abstraction to be GPU compatible (#298).

Just a note: I attempted this while working on abstract operations and was unable to solve the problem. But we are using new versions of julia + cuda tools all the time so it is / will be worth revisiting

**Comment by glwagner** [comment]
> This is probably an edge case but I don't think setting diffusivity BCs when using a tuple of closures will actually work. Or rather, I'm not sure how to specify two different diffusivity BCs, one for each closure.

For tuples of closures, everything related to the closure is tupled, including the `DiffusivityFields`. If there are different diffusivity fields for different closures, then it is possible to set them with different diffusivities?

**Comment by glwagner** [comment]
> For each turbulence closure, DiffusivityFields gets the same two versions (one if you want control over the full field, another if you just want to set boundary conditions).

I don't understand this point, can it be elaborated?

File: src/BoundaryConditions/field_boundary_conditions.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/BoundaryConditions/field_boundary_conditions.jl]: The `field_type` is used to call e.g. `default_tracer_bc` or `default_velocity_bc` in an `@eval` statement.

Yeah I wasn't super sure of the best way to define these but I did not want to define separate explicit constructors for each six of them as the function signature for each would be pretty complicated
```
function FieldBoundaryConditions(grid::AbstractGrid; field_type,
                                 west=default_bc(grid, field_type, 1), east=default_bc(grid, field_type, 1),
                                 south=default_bc(grid, field_type, 2), north=default_bc(grid, field_type, 2),
                                 bottom=default_bc(grid, field_type, 3), top=default_bc(grid, field_type, 3))
```
so I added a `field_type` kwarg and was able to write a single `FieldBoundaryConditions` function that handled all six field types.

I don't think harder to maintain (there's less code and really only one function that does anything), and I'm not sure what is not transparent about this piece of code.

Happy to change it, just not sure what improvements to make.

File: src/Fields/field.jl (hunk 2 above)
Comment (ali-ramadhan) [reply src/Fields/field.jl]: Yeah I was thinking of doing that but it's not strictly correct so I decided against it. A `CellField` could require `TracerBoundaryConditions`, `PressureBoundaryConditions`, or `DiffusivityBoundaryConditions` (right now they're all the same but that might not be true in the future). Even an `XFaceField` could be a velocity field or a field holding some flux.

I'm happy to extend the constructors for convenience. We can always refactor them when a future design change requires it.

File: src/Fields/field_tuples.jl (hunk 3 above)
Comment (ali-ramadhan) [reply src/Fields/field_tuples.jl]: I added an example to the docstring.

File: src/Fields/field_tuples.jl (hunk 4 above)
Comment (ali-ramadhan) [reply src/Fields/field_tuples.jl]: I've now defined u, v, w in the function signature and got rid of the ternary conditionals.

File: src/Models/incompressible_model.jl (hunk 5 above)
Comment (ali-ramadhan) [reply src/Models/incompressible_model.jl]: Sorry I should have been clearer what the model constructor does. Passing the empty `NamedTuple()` will impose default boundary conditions on every field as the `boundary_conditions` are passed to all the field tuple constructors.

Users now impose non-default boundary conditions by passing a named tuple of `FieldBoundaryConditions` where the keys are field names.

Some examples:
```julia
boundary_conditions = (u=u_bcs,)
boundary_conditions = (u=u_bcs, T=T_bcs)
boundary_conditions = (u=u_bcs, T=T_bcs, κₑ=(T=κₑ_bcs, S=κₑ_bcs))
```

An alternative API design could be a more hierarchical but more verbose named tuple (although I'm not a super huge fan of the verbosity here unless you want to call your tracers u, v, w, or κₑ for some reason):
```julia
boundary_conditions = (velocities=(u=u_bcs,),)
boundary_conditions = (velocities=(u=u_bcs,), tracers=(T=T_bcs,))
boundary_conditions = (velocities=(u=u_bcs,), tracers=(T=T_bcs,), diffusivities=(T=κₑ_bcs, S=κₑ_bcs))
```

File: src/TimeSteppers/TimeSteppers.jl (hunk 6 above)
Comment (ali-ramadhan) [reply src/TimeSteppers/TimeSteppers.jl]: It's still necessary as it's passed to `solve_for_pressure!` a few lines below.

File: src/TimeSteppers/TimeSteppers.jl (hunk 7 above)
Comment (ali-ramadhan) [reply src/TimeSteppers/TimeSteppers.jl]: Yeah this is one of the downsides of not having a fully functional field abstraction...

They are not the same because `tendencies` is a `datatuple` while `model.timestepper.Gⁿ` is a named tuple of fields. `fill_halo_regions` now requires fields so some functions want a `datatuple` while others want the latter.

I agree we should not be mixing the two as it's very confusing.

I think if we are committing to developing a functional field abstraction then we should keep the code as is as parts of it, such as the halo filling, already switched to using fields and has benefited, but put explicit comments anywhere that mixes the two explaining what is going on.

Then once we have a functional field abstraction we can get rid of all the comments and datatuples and we won't be mixing data tuples and field tuples anymore.

File: src/TurbulenceClosures/diffusivity_fields.jl (hunk 8 above)
Comment (ali-ramadhan) [reply src/TurbulenceClosures/diffusivity_fields.jl]: Done.

**Comment by ali-ramadhan** [comment]
> We should always use default boundary conditions for `grid` and the location, eg `Face` or `Cell` along each dimension.

I'm okay with doing this but should we be worried that not all `{Cell,Cell,Cell}` fields are tracer fields and not all `{Face,Cell,Cell}` fields are u-velocity fields?

> Just a note: I attempted this while working on abstract operations and was unable to solve the problem. But we are using new versions of julia + cuda tools all the time so it is / will be worth revisiting

Hmmm, do you remember what you got stuck on? I'd be in favor of reopening the issue and trying to adapt `Field` to work in GPU kernels.

> For tuples of closures, everything related to the closure is tupled, including the DiffusivityFields. If there are different diffusivity fields for different closures, then it is possible to set them with different diffusivities?

Ah sorry I meant that passing a named tuple like
```julia
boundary_conditions = (b=buoyancy_bcs, κₑ=(b=κₑ_bcs,))
```
to the model constructor will apply diffusivity boundary conditions as expected. But if you have two closures and you want to impose different diffusivity boundary conditions on each one then you can't do so by passing a named tuple to the model constructor. You can of course use `DiffusivityFields` to accomplish this.

We could upgrade the logic for interpreting boundary condition named tuples so that you can do something like
```julia
κ1_bcs = (T=some_diffusivity_bcs, S=some_diffusivity_bcs)
κ2_bcs = (T=some_other_diffusivity_bcs, S=some_other_diffusivity_bcs)
boundary_conditions = (T=T_bcs, κₑ=(κ1_bcs, κ2_bcs))
```
but it felt like a pretty edge case (very very few users would probably use it) so I didn't do it. But we can do it, it's within scope for this PR.

**Comment by ali-ramadhan** [comment]
> > For each turbulence closure, DiffusivityFields gets the same two versions (one if you want control over the full field, another if you just want to set boundary conditions).
> 
> I don't understand this point, can it be elaborated?

Ah the first one allows you to pass pre-constructed fields as keyword arguments. Users will probably not use this, it's mostly useful for custom diffusivity field tuple constructions. I had `restore_from_checkpoint` in mind when I wrote this constructor.

```julia
function TracerDiffusivityFields(arch, grid, tracer_names; kwargs...)
    κ_fields =
        Tuple(c ∈ keys(kwargs) ?
              kwargs[c] :
              CellField(arch, grid, DiffusivityBoundaryConditions(grid), zeros(arch, grid))
              for c in tracer_names)
    return NamedTuple{tracer_names}(κ_fields)
end
```

The second one allows you to pass non-default boundary conditions as a named tuple. Here `bcs` will be the named tuple users pass to the model constructor. Users will probably not use constructor either. This constructor is meant to be used by model constructors.

```julia
function TracerDiffusivityFields(arch, grid, tracer_names, bcs::NamedTuple)
    κ_fields =
        Tuple(c ∈ keys(bcs) ?
              CellField(arch, grid, bcs[c],                              zeros(arch, grid)) :
              CellField(arch, grid, DiffusivityBoundaryConditions(grid), zeros(arch, grid))
              for c in tracer_names)
    return NamedTuple{tracer_names}(κ_fields)
end
```

It might be possible to merge the two constructors (which have different uses) into one but I couldn't figure out how to do it. `VelocityFields` and `TracerFields` has two constructors for the same reason. `TendencyFields` only has one because users cannot specify tendency boundary conditions by passing a named tuple of `FieldBoundaryCondition`s to a model constructor.

**Comment by ali-ramadhan** [comment]
Another thing to potentially discuss if whether we want to switch from using `Face` and `Cell` to `Face` and `Center`. I think we agreed to do this at some point, this PR might be a good place to do so.

File: src/BoundaryConditions/field_boundary_conditions.jl (hunk 1 above)
Comment (glwagner) [reply src/BoundaryConditions/field_boundary_conditions.jl]: Here's what I propose:

1. We define constructors for boundary conditions that take `location`, `topology` , and optional specified pairs in each direction. The defaults are

```julia
DefaultBoundaryCondition(loc, ::Union{Periodic, Flat}) = PeriodicBC()

DefaultBoundaryCondition(Cell, ::Bounded) = NoFluxBC()

DefaultBoundaryCondition(Face, ::Bounded) = NoPenetrationBC()
```

2. `FieldBoundaryConditions` takes a location tuple (eg `(Face, Cell, Cell)`) as a required positional argument (rather than `field_type`), and then creates 3 pairs of `CoordinateBoundaryConditions` for each direction, for example, 

```julia
function FieldBoundaryConditions(grid, loc; 
    east = DefaultBoundaryCondition(topology(grid)[1], loc[1]),
    west = DefaultBoundaryCondition(topology(grid)[1], loc[1]),
    south = DefaultBoundaryCondition(topology(grid)[2], loc[2]),
    north = DefaultBoundaryCondition(topology(grid)[2], loc[2]),
    bottom = DefaultBoundaryCondition(topology(grid)[3], loc[3]),
    top = DefaultBoundaryCondition(topology(grid)[3], loc[3])
)

# validate the boundary conditions

    x = CoordinateBoundaryConditions(west, east)
    y = CoordinateBoundaryConditions(south, north)
    z = CoordinateBoundaryConditions(bottom, top)

    return FieldBoundaryConditions(x, y, z)
end
```

3. The convenience constructors call `FieldBoundaryConditions` with appropriate arguments; for example:

```julia
UVelocityBoundaryConditions(grid; user_defined_bcs...) = FieldBoundaryConditions(grid, (Face, Cell, Cell); user_defined_bcs...)
```

This will allow users to define fields with "default" boundary conditions for fields that are not of any `field_type`. For example, vertical vorticity is located at `(Face, Face, Cell)`. Users can still avoid having to remember "where" a field is on the grid by using our convenience constructors. I also think this may (?) shorten and simplify the code and reduce shenanigans (maybe not desirable since we are `Oceananigans`...?) We still want to perform validation so some shenanigans are inevitable.

Does that make sense? Any issues I'm not forseeing?

File: src/Fields/field.jl (hunk 2 above)
Comment (glwagner) [reply src/Fields/field.jl]: I don't think there is a correctness problem here. We provide a sensible default for convenience; and in the case that the default is incorrect, non-default, correct boundary conditions can be specified. For example, `zeros(10)` produces an array of `Float64`, which is often what users want and is therefore convenient. This is not *always* what users want, of course, so it is possible to write `zeros(Float32, 10)` if you need to. 

Note that tracers, pressures, and diffusivities do in fact share boundary conditions (right?) so the default is correct for all of them. I think it could be annoying to have to specify boundary conditions on a field that is, for example, intended only as scratch space for diagnostics calculations. For that application having a default is nice.

File: src/Models/incompressible_model.jl (hunk 5 above)
Comment (glwagner) [reply src/Models/incompressible_model.jl]: Yeah, I guess either way is fine. We could just write

```julia
boundary_conditions = (u = UVelocityBoundaryConditions(grid), v=VVelocityBoundaryConditions(grid), w=WVelocityBoundaryConditions(grid))
```

The only real purpose of doing that is code-as-documentation (if you're reading the model constructor, you can use the default as an example for what you need to do). Not sure if that's a good practice or not.

File: src/TimeSteppers/TimeSteppers.jl (hunk 6 above)
Comment (glwagner) [reply src/TimeSteppers/TimeSteppers.jl]: ah, I missed that. 

We could put the line

```julia
tendency_data = datatuple(model.timestepper.Gⁿ)
```

in this function and remove the tendencies argument. But maybe we should hold off since we might refactor time-stepping soon anyways?

File: src/TimeSteppers/TimeSteppers.jl (hunk 7 above)
Comment (glwagner) [reply src/TimeSteppers/TimeSteppers.jl]: Okay sorry, I didn't fully grasp the implications of putting boundary conditions into fields. So my comments are a bit silly about time stepping here.

One simplification could be to call `datatuple` locally. A single `datatuple` call was a sensible code simplification previously when boundary conditions and fields were divorced. Now that they are tied together I don't think that's the clearest way to write this code anymore.

We may also want to start appending data tuple variable names with `_data` for clarity too?

File: src/BoundaryConditions/field_boundary_conditions.jl (hunk 1 above)
Comment (glwagner) [reply src/BoundaryConditions/field_boundary_conditions.jl]: Just another thought: we can make the signatures defensive by adding `arch` as a required positional argument to `DefaultBoundaryCondition` so we are ready for distributed fields with `InteriorBoundaryCondition()` ?

**Comment by glwagner** [comment]
> I'm okay with doing this but should we be worried that not all {Cell,Cell,Cell} fields are tracer fields and not all {Face,Cell,Cell} fields are u-velocity fields?

Defaults are not about correctness, but convenience --- the concept of a "default" implies that it is not always correct (otherwise we would not use the word "default", because it would never have to change). I think defaults can make user's lives simpler. I think there may be a lot of cases in field definition where boundary conditions are unused.

> Hmmm, do you remember what you got stuck on?

Was just getting the dreaded "dynamic function invocation". It's worth a shot and documenting our progress. It seems unlikely there's a fundamental issue. We just have to figure it out.

> But if you have two closures and you want to impose different diffusivity boundary conditions on each one then you can't do so by passing a named tuple to the model constructor. You can of course use DiffusivityFields to accomplish this.

Okay thank you I understand. The issue here is the ability to use defaults. For that, we would have to require users to *name* closures when used in a tuple. 

I am happy to leave this functionality as "unsupported". That said, I suspect we will have to confront this issue in the future if we want to do sophisticated GCM stuff, which can have complicated sets of turbulence closures.

**Comment by glwagner** [comment]
> but it felt like a pretty edge case (very very few users would probably use it) so I didn't do it.

It may become common if users do realistic GCM stuff, because there will be many turbulence closures, but only *one* of them would require explicit boundary-diffusivity modeling to control field gradients and diffusivities along boundaries (for example, prescribing some interesting lateral diffusivity parameterization on top of a vertical mixing parameterization).

File: src/BoundaryConditions/field_boundary_conditions.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/BoundaryConditions/field_boundary_conditions.jl]: Thanks for the suggestions, looks great! Just updated the code, I do think it's simpler now.

Comment (ali-ramadhan) [reply src/BoundaryConditions/field_boundary_conditions.jl]: > Just another thought: we can make the signatures defensive by adding `arch` as a required positional argument to `DefaultBoundaryCondition` so we are ready for distributed fields with `InteriorBoundaryCondition()` ?

Hmmm, I thought about it a bit but am a little hesitant as I'm not sure what a distributed field might want. What I do in https://github.com/climate-machine/Oceananigans.jl/pull/590 is "inject halo communication boundary conditions" after the user has specified any non-default BCs so `arch` isn't necessarily needed but of course it's still being designed. There are also no distributed fields, just distributed models (a wrapper around a model plus MPI stuff like rank connectivity).

Should be an easy enough change to make when/if we need it.

File: src/Fields/field.jl (hunk 2 above)
Comment (ali-ramadhan) [reply src/Fields/field.jl]: That's a good point. Will change.

File: src/Models/incompressible_model.jl (hunk 5 above)
Comment (ali-ramadhan) [reply src/Models/incompressible_model.jl]: > Of course if non-defaults are used for some fields then we still have to infer defaults for the rest of the fields so we don't really help ourselves practically with a default.

True but the examples and model setup docs have plenty of examples of non-default boundary conditions so I don't think it's a bad idea to show an actual named tuple in the docstring as opposed to an empty `NamedTuple()` which you'd almost never use as a user.

File: src/TimeSteppers/TimeSteppers.jl (hunk 6 above)
Comment (ali-ramadhan) [reply src/TimeSteppers/TimeSteppers.jl]: Hmmm, yeah we'll probably be the next ones to refactor the time stepping code so might as well leave it as all the tests pass.

File: src/TimeSteppers/TimeSteppers.jl (hunk 7 above)
Comment (ali-ramadhan) [reply src/TimeSteppers/TimeSteppers.jl]: I like the idea of using `_data` affixes but hopefully we try and flesh out a full field abstraction then this won't be necessary. But if it doesn't work out, then we should use this.

**Comment by ali-ramadhan** [comment]
> It may become common if users do realistic GCM stuff, because there will be many turbulence closures, but only *one* of them would require explicit boundary-diffusivity modeling to control field gradients and diffusivities along boundaries (for example, prescribing some interesting lateral diffusivity parameterization on top of a vertical mixing parameterization).

**Comment by ali-ramadhan** [comment]
> It may become common if users do realistic GCM stuff, because there will be many turbulence closures, but only *one* of them would require explicit boundary-diffusivity modeling to control field gradients and diffusivities along boundaries (for example, prescribing some interesting lateral diffusivity parameterization on top of a vertical mixing parameterization).

Ah interesting I didn't think of such cases. I believe we can set up arbitrarily complicated boundary conditions so anything is possible. It's just hard and probably unfeasible/undesirable to extend the user interface to handle all these cases.

File: src/BoundaryConditions/field_boundary_conditions.jl (hunk 1 above)
Comment (glwagner) [reply src/BoundaryConditions/field_boundary_conditions.jl]: Fair enough.

**Comment by ali-ramadhan** [comment]
> Thought: we provide the sugary syntax `.top` and `.bottom` for boundary conditions in `z`. Should we also provide east, west, south, and north, and avoid using `.left` and `.right` in the code for full clarity?

Sounds like a good idea. I think `setbc!` and `getbc` aren't fully tested so also worth adding some simple tests there.

I'll open an issue to document this.
